packages/scenario/src/jr/Scenario.ts

```diff
@@ -288,13 +294,73 @@ export class Scenario {
 		return this.setNonTerminalEventPosition(increment, expectReference);
 	}
 
-	private setPositionalStateToReference(reference: string): AnyPositionalEvent {
+	private createMissingRepeatInstances(reference: string): void {
+		let tempReference = reference;
+		let indexedReference: string | null = null;
+
+		const trailingPositionalPredicatePattern = /\[\d+\]$/;
+
+		do {
+			if (trailingPositionalPredicatePattern.test(tempReference)) {
+				indexedReference = tempReference;
+			} else {
+				tempReference = tempReference.replace(/\/[^/]+$/, '');
+
+				if (tempReference === '') {
+					break;
+				}
+			}
+		} while (indexedReference == null);
+
+		if (indexedReference == null) {
+			return;
+		}
+
+		const repeatRangeReference = indexedReference.replace(trailingPositionalPredicatePattern, '');
+
+		const positionalPredicate = indexedReference.replace(/^.*\[(\d+)\]$/, '$1');
+		const count = parseInt(positionalPredicate, 10);
+
+		if (count < 1) {
+			return;
+		}
+
+		const repeatRange = this.getInstanceNode(repeatRangeReference);
+
+		if (repeatRange.nodeType !== 'repeat-range') {
+			throw 'todo';
+		}
+
+		const repeatBodyDefinition = repeatRange.definition.bodyElement;
+
+		if (repeatBodyDefinition.countExpression != null || repeatBodyDefinition.isFixedCount) {
+			return;
+		}
+
+		const instances = repeatRange.currentState.children.length;
+		const delta = count - instances;
+
+		for (let i = 0; i < delta; i += 1) {
+			this.createNewRepeat(repeatRangeReference);
+		}
```

Comment:
The comment about building blocks is partially addressed at https://github.com/getodk/web-forms/pull/150/commits/66aa8c979c655adaf461807266fb58c66acee8e6#diff-f3ac6c9869393022cd23375d383ded81f1d35c2c962125846cb288792ee28d52R15

Reply:
I've been mulling this since I saw these comments yesterday, and my instinct is that we're probably better off not porting this aspect of `Scenario`.

- The tests I've encountered which assume this behavior have frequently surprised and confused me. And not in a trivial way: almost every time, I've misinterpreted the nature of a failure, or misunderstood the intent of the test, sending me off on unproductive tangents.

- Tests which expect this level of logical complexity should probably have their own logic tested as well. For instance, I was already deeply uncomfortable with the level of complexity in `child-vaccination`. There is significant application-like work happening there, and I find it vastly harder to understand the test flow than any of the tests which more explicitly express what they are doing.

- I should say that I regret the linked comment. In hindsight, the kind of expression analysis and manipulation that I discuss there, and that you say you were expecting here, is definitely appropriate for _engine implementation details_. I am quite cautious about using logic of that kind to implement _test behavior_.

- On Tuesday we discussed the fact that such logic is in fact incoming, supporting a variety of other **engine improvements**. And as we discussed then, that logic is downstream from this change, as other affected tests would currently depend on this specific implicit repeat instance creation. But to be clear, as that work took form, it felt pretty clear that the expression analysis/manipulation logic will be entirely encapsulated in the engine. If we do anticipate exposing arbitrarily/increasingly complex expression analysis and manipulation as a public API, I think we should really think hard about what the appropriate design for that would be. And the appropriate package and/or abstraction layer from which we should expose it.

I already wanted to make repeat instance creation more explicit before this discussion, but I hesitated in this PR on the basis that it would increase divergence from JavaRosa. But that divergence is much more appealing to me now, in light of some of the implications this has raised.

Reply:
I thought your linked comment meant in the engine. Agreed it’s the engine that eventually should have that functionality so good to see it’s coming. 👍